File: 59.py
from collections import defaultdict, deque

def alienOrder(words):
    # Step 1: Initialize graph
    graph = defaultdict(set)
    indegree = {c: 0 for w in words for c in w}

    # Step 2: Build graph by comparing adjacent words
    for i in range(len(words) - 1):
        w1, w2 = words[i], words[i + 1]
        # Invalid case: prefix longer but comes before shorter
        if len(w1) > len(w2) and w1.startswith(w2):
            return ""
        for c1, c2 in zip(w1, w2):
            if c1 != c2:
                if c2 not in graph[c1]:
                    graph[c1].add(c2)
                    indegree[c2] += 1
                break

    # Step 3: BFS (Topological Sort)
    q = deque([c for c in indegree if indegree[c] == 0])
    res = []

    while q:
        c = q.popleft()
        res.append(c)
        for nei in graph[c]:
            indegree[nei] -= 1
            if indegree[nei] == 0:
                q.append(nei)

    return "".join(res) if len(res) == len(indegree) else ""

File: 59.js
function alienOrder(words) {
  const graph = new Map();
  const indegree = new Map();

  // Initialize all characters
  for (const w of words)
    for (const c of w) indegree.set(c, 0);

  // Build graph
  for (let i = 0; i < words.length - 1; i++) {
    const w1 = words[i], w2 = words[i + 1];
    if (w1.length > w2.length && w1.startsWith(w2)) return "";

    for (let j = 0; j < Math.min(w1.length, w2.length); j++) {
      if (w1[j] !== w2[j]) {
        if (!graph.has(w1[j])) graph.set(w1[j], new Set());
        if (!graph.get(w1[j]).has(w2[j])) {
          graph.get(w1[j]).add(w2[j]);
          indegree.set(w2[j], indegree.get(w2[j]) + 1);
        }
        break;
      }
    }
  }

  // Kahn’s BFS
  const queue = [];
  for (const [ch, deg] of indegree)
    if (deg === 0) queue.push(ch);

  const result = [];
  while (queue.length) {
    const c = queue.shift();
    result.push(c);
    if (graph.has(c)) {
      for (const nei of graph.get(c)) {
        indegree.set(nei, indegree.get(nei) - 1);
        if (indegree.get(nei) === 0) queue.push(nei);
      }
    }
  }

  return result.length === indegree.size ? result.join('') : "";
}
